[Streams] Update Streams IA and structure. (#7062)
elastic/docs-content: solutions/observability/streams-new/get-data-in.md

Best for new deployments, custom logs, and mixed-format sources.
- **[Work with existing data](#get-data-in-classic)**: Use classic streams to work with data already flowing into {{es}}. No migration or configuration changes required.

## Before you get started [get-data-in-prerequisites]

Streams requires the following permissions:

::::{applies-switch}

:::{applies-item} serverless:
Streams requires these {{serverless-full}} roles:

- Admin: Ability to manage all Streams
- Editor/Viewer: Limited access, cannot perform all actions

:::

:::{applies-item} stack:
To manage all streams, you need the following permissions:

- **Cluster permissions**: `manage_index_templates`, `manage_ingest_pipelines`, `manage_pipeline`, `read_pipeline`
- **Data stream level permissions**: `read`, `write`, `create`, `manage`, `monitor`, `manage_data_stream_lifecycle`, `read_failure_store`, `manage_failure_store`, `manage_ilm`.

To view streams, you need the following permissions:
- **Data stream level**: `read`, `view_index_metadata`, `monitor`

> For more information, refer to [Cluster privileges](elasticsearch://reference/elasticsearch/security-privileges.md#privileges-list-cluster) and [Granting privileges for data streams and aliases](../../../deploy-manage/users-roles/cluster-or-deployment-auth/granting-privileges-for-data-streams-aliases.md)

Reviewer: ```suggestion
For more information, refer to [Cluster privileges](elasticsearch://reference/elasticsearch/security-privileges.md#privileges-list-cluster) and [Granting privileges for data streams and aliases](../../../deploy-manage/users-roles/cluster-or-deployment-auth/granting-privileges-for-data-streams-aliases.md).
```

:::

::::

## Ingest new data [get-data-in-wired]

```{applies_to}
stack: preview 9.2+
serverless: preview
```

Wired streams send your documents to a managed endpoint, from which you can route data into child streams based on [partitioning](./organize-your-data.md) rules. Child streams automatically inherit mappings, lifecycle settings, and processors from the parent, and configuration changes propagate through the hierarchy.

> {applies_to}`stack: preview 9.4+` {applies_to}`serverless: preview` Two endpoints are available. Use **`logs.otel`** (recommended) when sending OTel-native data or when you want a consistent, normalized format — Streams translates ECS field names to OTel equivalents automatically. Use **`logs.ecs`** when your data already uses ECS field names and you want to preserve them without transformation.

Reviewer: Reorder `applies_to` tags to start with Serverless - also suggesting to use a list here for easier reading:

```suggestion
{applies_to}`serverless: preview` {applies_to}`stack: preview 9.4+` Two endpoints are available:
- Use **`logs.otel`** (recommended) when sending OTel-native data or when you want a consistent, normalized format — Streams translates ECS field names to OTel equivalents automatically.
- Use **`logs.ecs`** when your data already uses ECS field names and you want to preserve them without transformation.
```

Alternative syntax (takes care of ordering automatically if you start with stack first):

```suggestion
{applies_to}`{ serverless: preview, stack: preview 9.4+ }` Two endpoints are available:
- Use **`logs.otel`** (recommended) when sending OTel-native data or when you want a consistent, normalized format — Streams translates ECS field names to OTel equivalents automatically.
- Use **`logs.ecs`** when your data already uses ECS field names and you want to preserve them without transformation.
```

> {applies_to}`{serverless: preview, stack: preview 9.4+}` Two endpoints are available. Use **`logs.otel`** (recommended) when sending OTel-native data or when you want a consistent, normalized format — Streams translates ECS field names to OTel equivalents automatically. Use **`logs.ecs`** when your data already uses ECS field names and you want to preserve them without transformation.

Reviewer 2: > Streams translates ECS field names to OTel equivalents automatically

I could be wrong here but I don't think this is the case - where did you get this from @mdbirnstiehl ?

Reviewer 3: @3kt  This was pulled over from the old docs that say:

When data is ingested into a wired stream, it’s automatically translated into this normalized format:
- Standard ECS documents are converted to OTel fields (`message → body.text`, `log.level → severity_text`, `host.name → resource.attributes.host.name`, and so on).
- Custom fields are stored under `attributes.*`.

Maybe this isn't correct or I misinterpreted here?

Reviewer 2: you're right - see this [thread](https://elastic.slack.com/archives/C0BDCUT2DEW/p1784214028722699)

To send data to a wired stream, configure your shipper to point to the appropriate endpoint:

:::::{tab-set}

::::{tab-item} OpenTelemetry
:::{note}
Set the index based on your {{stack}} version:

> - {applies_to}`serverless: preview` {applies_to}`stack: preview 9.4+` Set the index to `logs.otel` or `logs.ecs`, depending on which endpoint you want to use.

Reviewer: Maybe?
```suggestion
- {applies_to}`serverless: preview` {applies_to}`stack: preview 9.4+` Set the index to `logs.otel` or `logs.ecs`, depending on the endpoint you want to use.
```

- {applies_to}`stack: preview 9.2-9.3` Set the index to `logs`. Only the `logs` endpoint is available in these versions.
:::

```yaml
processors:
  transform/logs-streams:
    log_statements:
      - context: resource
        statements:

>           - set(attributes["elasticsearch.index"], "logs.otel") # Set to `logs.otel` or `logs.ecs` (serverless and stack 9.4+), or `logs` (stack 9.2–9.3)

Reviewer: ```suggestion
          - set(attributes["elasticsearch.index"], "logs.otel") # Set to `logs.otel` or `logs.ecs` (Serverless and Stack 9.4+), or `logs` (Stack 9.2–9.3).
```

service:
  pipelines:
    logs:

>       receivers: [myreceiver] # works with any logs receiver

Reviewer: ```suggestion
      receivers: [myreceiver] # Works with any logs receiver.
```

      processors: [transform/logs-streams]

>       exporters: [elasticsearch, otlp] # works with either

Reviewer: ```suggestion
      exporters: [elasticsearch, otlp] # Works with either exporter.
```

```
::::

::::{tab-item} Filebeat
:::{note}
Set the index based on your {{stack}} version:

- {applies_to}`stack: preview 9.2-9.3` Set the index to `logs`. Only the `logs` endpoint is available in these versions.

> - {applies_to}`serverless: preview` {applies_to}`stack: preview 9.4+` Set the index to `logs.otel` or `logs.ecs`, depending on which endpoint you want to use.

Reviewer: Reorder the list to start with Serverless and the latest Stack version:

```suggestion
- {applies_to}`serverless: preview` {applies_to}`stack: preview 9.4+` Set the index to `logs.otel` or `logs.ecs`, depending on the endpoint you want to use.
- {applies_to}`stack: preview 9.2-9.3` Set the index to `logs`. Only the `logs` endpoint is available in these versions.
```

:::

```yaml
filebeat.inputs:
  - type: filestream
    id: my-filestream-id

>     index: logs.otel # Set to `logs.otel` or `logs.ecs` (serverless and stack 9.4+), or logs (stack 9.2–9.3)

Reviewer: ```suggestion
    index: logs.otel # Set to `logs.otel` or `logs.ecs` (Serverless and Stack 9.4+), or `logs` (Stack 9.2–9.3).
```

    enabled: true
    paths:
      - /var/log/*.log

# No need to install templates for wired streams
setup:
  template:
    enabled: false

output.elasticsearch:
  hosts: ["<elasticsearch-host>"]
  api_key: "<your-api-key>"
```
::::

::::{tab-item} Logstash
:::{note}
Set the index based on your {{stack}} version:

- {applies_to}`stack: preview 9.2-9.3` Set the index to `logs`. Only the `logs` endpoint is available in these versions.

> - {applies_to}`serverless: preview` {applies_to}`stack: preview 9.4+` Set the index to `logs.otel` or `logs.ecs`, depending on which endpoint you want to use.

Reviewer: Reorder the list to start with Serverless and the latest Stack version:

```suggestion
- {applies_to}`serverless: preview` {applies_to}`stack: preview 9.4+` Set the index to `logs.otel` or `logs.ecs`, depending on the endpoint you want to use.
- {applies_to}`stack: preview 9.2-9.3` Set the index to `logs`. Only the `logs` endpoint is available in these versions.
```

:::

```json
output {
  elasticsearch {
    hosts => ["<elasticsearch-host>"]
    api_key => "<your-api-key>"

>     index => "logs.otel" # Set to `logs.otel` or `logs.ecs` (serverless and stack 9.4+), or `logs` (stack 9.2–9.3)

Reviewer: ```suggestion
    index => "logs.otel" # Set to `logs.otel` or `logs.ecs` (Serverless and Stack 9.4+), or `logs` (Stack 9.2–9.3).
```

    action => "create"
  }
}
```
::::

::::{tab-item} Fleet
Use the **Custom Logs (Filestream)** integration to send data to wired streams:

1. Find **{{fleet}}** in the navigation menu or use the [global search field](../../../explore-analyze/find-and-organize/find-apps-and-objects.md).
1. Select the **Settings** tab.
1. Under **Outputs**, find the output you want to use and select the {icon}`pencil` icon.
1. Turn on **Write to logs streams**.
1. Add the **Custom Logs (Filestream)** integration to an agent policy.
1. Enable the **Use the "logs" data stream** setting under **Change defaults**.
1. Under **Where to add this integration**, select an agent policy that uses the output configured in step 4.
::::

::::{tab-item} API
:::{note}
Set the endpoint based on your {{stack}} version:

- {applies_to}`stack: preview 9.2-9.3` Set the endpoint to `logs`. Only the `logs` endpoint is available in these versions.

> - {applies_to}`serverless: preview` {applies_to}`stack: preview 9.4+` Set the endpoint to `logs.otel` or `logs.ecs`, depending on which endpoint you want to use.

Reviewer: Reorder the list to start with Serverless and the latest Stack version:

```suggestion
- {applies_to}`serverless: preview` {applies_to}`stack: preview 9.4+` Set the endpoint to `logs.otel` or `logs.ecs`, depending on the endpoint you want to use.
- {applies_to}`stack: preview 9.2-9.3` Set the endpoint to `logs`. Only the `logs` endpoint is available in these versions.
```

:::

Send data to the endpoint using the [Bulk API]({{es-apis}}operation/operation-bulk):

```json

> POST /logs.otel/_bulk # Set to `logs.otel` or `logs.ecs` (serverless or stack 9.4+), or `logs` (stack 9.2–9.3)

Reviewer: ```suggestion
POST /logs.otel/_bulk # Set to `logs.otel` or `logs.ecs` (Serverless or Stack 9.4+), or `logs` (Stack 9.2–9.3).
```

{ "create": {} }
{ "@timestamp": "2025-05-05T12:12:12", "message": "Hello world!", "host.name": "my-host-name" }
```
::::

:::::

### Verify data is flowing

After configuring your data source, confirm data is appearing in Discover.

For wired streams, you first need to make the index pattern available:

1. Manually [create a {{data-source}}](../../../explore-analyze/find-and-organize/data-views.md#settings-create-pattern) for the wired streams index pattern (`logs,logs.*`).
1. Add the wired streams index pattern (`logs,logs.*`) to the `observability:logSources` {{kib}} advanced setting, which you can open from the navigation menu or by using the [global search field](../../../explore-analyze/find-and-organize/find-apps-and-objects.md).

Once data appears in Discover, you're ready to start organizing, parsing, and configuring retention for your streams.

## Work with existing data [get-data-in-classic]

Classic streams let you use the Streams UI to extract fields and configure data retention for data that's already being ingested into {{es}} without additional configuration.

:::{warning}

> Do not reroute or migrate existing data streams to the `logs`, `logs.otel`, or `logs.ecs` endpoints. Classic streams work with your existing data streams in place — no migration is required or recommended.

Reviewer 2: ```suggestion
Do not reroute or migrate existing data streams to the `logs`, `logs.otel`, or `logs.ecs` endpoints. Classic streams work with your existing data streams in place — no migration is required.
```

:::

Classic streams:

- Are based on existing data streams, index templates, and component templates.
- Can follow the data retention policy set in the existing index template.
- Do not support hierarchical inheritance or cascading configuration updates.

Open classic streams from the following places in {{kib}}:

- Select **Streams** from the navigation menu or use the [global search field](../../../explore-analyze/find-and-organize/find-apps-and-objects.md), and find the data stream in the Streams table.

- Open the data stream for a specific document from **Discover**. To do this, expand the details flyout for a document stored in a data stream, and select **Stream** or an action associated with the document's data stream. Streams then opens filtered to the selected data stream.
    1. From the **Documents** table, select the {icon}`expand` icon for a document to expand the details flyout.

>     1. Under **Stream**, select the link to the stream (for example, `logs-generic.default`). The Streams UI opens filtered to the stream that contains the document.

Reviewer: This seems like a duplication, we should keep either the two sentences before the steps, or the steps.

> You can also access Streams features using the Streams API. {applies_to}`stack: preview 9.1+` {applies_to}`serverless: preview` Refer to the [Streams API documentation]({{kib-apis}}group/endpoint-streams) for more information.

Reviewer: I think this would be a better placement of the `applies_to` tags:

```suggestion
{applies_to}`serverless: preview` {applies_to}`stack: preview 9.1+` You can also access Streams features using the Streams API. Refer to the [Streams API documentation]({{kib-apis}}group/endpoint-streams) for more information.
```
